compiler/util/misc.cpp

```diff
@@ -73,8 +73,14 @@ bool requireWideReferences() {
   return !fLocal || forceWidePtrs();
 }
 
+//
+// If the --no-local flag is used or the locale model is not 'flat'
+// (i.e., has sub-locales that on-clauses might target), we should
+// not drop on-clauses on the floor.  I.e., we should require them
+// to be "outlined."
+//
 bool requireOutlinedOn() {
-  return !fLocal || forceWidePtrs();
+  return !fLocal || strcmp(CHPL_LOCALE_MODEL, "flat");
```

Comment:
I've pushed a commit that incorporates your suggestions and am re-testing if you want to take a look.